src/SkipStages.cpp

```diff
@@ -206,16 +204,57 @@ class PredicateFinder : public IRVisitor {
 
 class ProductionGuarder : public IRMutator {
 public:
-    ProductionGuarder(const string &b, Expr p): buffer(b), predicate(p) {}
+    ProductionGuarder(const string &b, Expr compute_p, Expr alloc_p):
+        buffer(b), compute_predicate(compute_p), alloc_predicate(alloc_p) {}
 private:
     string buffer;
-    Expr predicate;
+    Expr compute_predicate;
+    Expr alloc_predicate;
 
     using IRMutator::visit;
 
+    void visit(const LetStmt *op) {
+        IRMutator::visit(op);
+
+        op = stmt.as<LetStmt>();
+        internal_assert(op);
+
+        const Call *c = op->value.as<Call>();
+        // We need to guard call to halide_memoization_cache_lookup to only be executed
+        // if the corresponding buffer is allocated.
+        if (c && (c->name == "halide_memoization_cache_lookup")) {
```

Comment:
I think both of these should be in the Call visitor instead, and should also check that this is memoizing the same buffer as the one being guarded. There may be multiple nested memoized Funcs going on.